modules/avfilter/util.c

```diff
@@ -0,0 +1,92 @@
+/**
+ * @file util.c	 Video filter using libavfilter -- utility functions
+ *
+ * Copyright (C) 2020 Mikhail Kurkov
+ */
+
+#include <string.h>
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+#include <libavutil/frame.h>
+#include "util.h"
+
+static int swap_lines(uint8_t *a, uint8_t *b, size_t size) {
+	uint8_t tmp[size];
+	memcpy(tmp, a, size);
+	memcpy(a, b, size);
+	memcpy(b, tmp, size);
+	return 0;
+}
+
+static int reverse_lines(uint8_t *data, int linesize, int count) {
+	for (int i = 0; i < count/2; i++)
+		swap_lines(data + linesize * i,
+			   data + linesize * (count - i - 1),
+			   abs(linesize));
+	return 0;
+}
+
+/*
+ * Sometimes AVFrame contains planes with lines in bottom-up order.
+ * Then linesize is negative and data points to the last row in buffer.
```

Comment:
this sounds strange, I have never seen a negative linesize. Which plane was it (0 to 3) ?

the linesize is the number of pixels in a horizontal line, for each plane.
example: [320, 160, 160, 0]

Reply:
For bottom-up frames all planes have negative linesizes, like this:  `linesize = {-1920, -960, -960, 0, 0, 0, 0, 0}` (this is from gdb inspecting AVFrame structure).